api

```diff
@@ -179,45 +183,39 @@ package_is_new_enough() { #check if the $1 package has an available version grea
   fi
 }
 
-anything_installed_from_repo() { #Given an apt repository URL, determine if any packages from it are currently installed
-  [ -z "$1" ] && error "anything_installed_from_repo: A repository URL must be specified."
+anything_installed_from_uri_suite_component() { #Given an apt repository uri, suite, and component, determine if any packages from it are currently installed
+  [ -z "$1" ] && error "anything_installed_from_uri_suite_component: A repository uri must be specified."
+  [ -z "$2" ] && error "anything_installed_from_uri_suite_component: A repository suite must be specified."
   
-  #user input repo-url. Remove 'https://', and translate '/' to '_' to conform to apt file-naming standard, with trailing _ to ensure full matches
-  local url="$(echo "$1" | sed 's+.*://++g' | tr '/' '_')_"
+  #component is an optional specification
+  if [ -z "$3" ]; then
+    local filepath="/var/lib/apt/lists/$(echo "$1" | sed 's+.*://++g' | sed "s,/$,," | tr '/' '_')_$(echo "$2" | sed "s,/$,," | tr '/' '_')_"
+  else
+    local filepath="/var/lib/apt/lists/$(echo "$1" | sed 's+.*://++g' | sed "s,/$,," | tr '/' '_')_dists_$(echo "$2" | sed "s,/$,," | tr '/' '_')_$(echo "$3" | sed "s,/$,," | tr '/' '_')_"
+  fi
+  debug $filepath
   
-  #find all package-lists pertaining to the url
-  local repofiles="$(ls /var/lib/apt/lists/*_Packages | grep -F "$url")"
+  #find all relevant package-lists
+  local repofiles="$(ls $filepath*Packages)"
+  debug "$repofiles"
   
   #for every repo-file, check if any of them have an installed file
-  local found=0
   local IFS=$'\n'
   local repofile
   for repofile in $repofiles ;do
     #search the repo-file for installed packages
+
+    local packages_in_repo="$(grep '^Package' "$repofile" | awk '{print $2}' | sort)"
+    local installed_packages="$(apt list --installed 2>/dev/null | tail -n +2 | awk -F/ '{print $1}')"
+    local apt_cache_policy_output="$(echo "$packages_in_repo" | list_intersect "$installed_packages" | tr '\n' ' ' | xargs -r apt-cache policy)"
```

Comment:
With tr used here, it seems to me like the output from this would always be a single line, meaning all matching packages would be passed to apt-cache in one run.
So it seems like the only use for xargs is shorthand to avoid running apt-cache if no packages from the repo are installed?

Reply:
apt-cache policy needs the input as individual arguments. it also does not read from stdin. these are the reasons why xargs is used.

Reply:
I explained poorly.
If apt-cache is only being run once, then something like this should work if I understand this correctly
```bash
local apt_cache_policy_output="$(apt-cache policy $(echo "$packages_in_repo" | list_intersect "$installed_packages" | tr '\n' ' '))"
```

Reply:
No I correctly interpreted what you said.

> apt-cache policy needs the input as individual arguments.

What you just sent puts all output as one input argument I think. Can't check now.

Reply:
No quotes around the inner `$()` will make each package into a separate argument.
If nothing else, this is simply something useful to know. I'm not necessarily requesting any changes here.